File: src/rffi/src/ref_count.cc
/*
 * Copyright 2019-2021 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

#include "rffi/api/ref_count.h"
#include "rtc_base/logging.h"

namespace webrtc {
namespace rffi {

// Decrements the ref count of a ref-counted object.
// If the ref count goes to zero, the object is deleted.
RUSTEXPORT void
Rust_decRc(rtc::RefCountInterface* owned_rc) {
  if (!owned_rc) {
    return;
  }

  auto result = owned_rc->Release();
  RTC_LOG(LS_VERBOSE) << "Did it get deleted? " << (result == rtc::RefCountReleaseStatus::kDroppedLastRef);
}

// Increments the ref count of a ref-counted object.
// The borrowed RC becomes an owned RC.
RUSTEXPORT void
Rust_incRc(rtc::RefCountInterface* borrowed_rc) {
  if (!borrowed_rc) {
    return;
  }

  borrowed_rc->AddRef();
}

}
}

File: src/rust/src/webrtc/sim/ref_count.rs
//
// Copyright 2019-2021 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

//! Simulation of Wrapper around rtc::RefCountInterface

use crate::webrtc;

/// # Safety
pub fn dec<T: webrtc::ptr::RefCounted>(_rc: webrtc::ptr::OwnedRc<T>) {
    info!("ref_count::dec()");
}

/// # Safety
pub unsafe fn inc<T: webrtc::ptr::RefCounted>(
    rc: webrtc::ptr::BorrowedRc<T>,
) -> webrtc::ptr::OwnedRc<T> {
    info!("ref_count::inc()");
    webrtc::ptr::OwnedRc::from_ptr(rc.as_ptr())
}
